arcgis-ios-sdk-samples/Display information/Filter by time extent/FilterByTimeExtentViewController.swift

```diff
@@ -0,0 +1,128 @@
+            // Set the map view's viewpoint.
+            let center = AGSPoint(x: -58.495293, y: 29.979774, spatialReference: .wgs84())
+            mapView.setViewpoint(AGSViewpoint(center: center, scale: 147_914_381.897889))
+        }
+    }
+    
+    /// The date formatter.
+    static let formatter: DateFormatter = {
+        let formatter = DateFormatter()
+        formatter.dateFormat = "yyyy/MM/dd HH:mm"
+        return formatter
+    }()
+    
+    /// The default date and time for the starting thumb.
+    let currentStartTime = formatter.date(from: "2005/10/01 05:00")!
+    /// The default date and time for the ending thumb.
+    let currentEndTime = formatter.date(from: "2005/10/31 05:00")!
+    
+    /// The feature layer tracking hurricanes in 2005.
+    let featureLayer = AGSFeatureLayer(
+        item: AGSPortalItem(
+            portal: .arcGISOnline(withLoginRequired: false),
+            itemID: "49925d814d7e40fb8fa64864ef62d55e"
+        ),
+        layerID: 0
+    )
+    
+    /// The time slider from the ArcGIS toolkit.
+    let timeSlider = TimeSlider()
+    
+    /// Configure the time slider's attributes and position.
+    func setupTimeSlider() {
+        // Configure time slider.
+        timeSlider.labelMode = .thumbs
+        timeSlider.addTarget(self, action: #selector(timeSliderValueChanged(timeSlider:)), for: .valueChanged)
+        // Add the time slider to the view.
+        view.addSubview(timeSlider)
+        
+        // Add constraints to position the slider.
+        let margin: CGFloat = 10.0
+        timeSlider.translatesAutoresizingMaskIntoConstraints = false
+        // Assign the constraints.
+        let constraints = [
+            timeSlider.bottomAnchor.constraint(equalTo: mapView.attributionTopAnchor, constant: -margin),
+            timeSlider.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: margin),
+            timeSlider.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -margin)
+        ]
+        // Activate the constraints.
+        NSLayoutConstraint.activate(constraints)
+    }
+    
+    /// Initialize the time slider's steps.
+    func initializeTimeStepsFromQuery() {
+        featureLayer.load { [unowned self] error in
+            guard error == nil else { return }
+            populateFeaturesWithQuery { timeExtent in
+                timeSlider.initializeTimeSteps(timeStepCount: 200, fullExtent: timeExtent) { _ in
+                    // Show the time slider.
+                    timeSlider.currentExtent = AGSTimeExtent(startTime: currentStartTime, endTime: currentEndTime)
+                    
+                    mapView.map!.operationalLayers.add(featureLayer)
+                }
+            }
+        }
+    }
+    
+    /// Populate the features using the requested time extent.
+    func populateFeaturesWithQuery(completion: @escaping (AGSTimeExtent) -> Void) {
```

Comment:
Sorry I didn't make myself clear. I don't think this is a must-have. The commit 6071de6085029e591ce82c333c73da3d19df6dde where we approved is already fine.

But if you want to keep it, below is what I'd suggest

1. Use the current time extent as query parameter, to only manually get a smaller amount of features for the initial display. When the time slider starts to play, more features will be shown as the slider progresses.
2. Adjust the time steps and playback interval, both to show off we can customize them in the time slider, and to increase the playback speed.

<details><summary>Code</summary>

```swift
// Copyright 2021 Esri
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import UIKit
import ArcGIS
import ArcGISToolkit

class FilterByTimeExtentViewController: UIViewController {
    /// The map view.
    @IBOutlet var mapView: AGSMapView! {
        didSet {
            mapView.map = AGSMap(basemapStyle: .arcGISTopographic)
            // Set the map view's viewpoint.
            let center = AGSPoint(x: -58.495293, y: 29.979774, spatialReference: .wgs84())
            mapView.setViewpoint(AGSViewpoint(center: center, scale: 1.5e8))
        }
    }
    
    /// The date formatter.
    static let formatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy/MM/dd HH:mm"
        return formatter
    }()
    
    /// The default date and time for the starting and ending thumb.
    let currentTimeExtent = AGSTimeExtent(
        startTime: formatter.date(from: "2005/10/01 05:00")!,
        endTime: formatter.date(from: "2005/10/31 05:00")!
    )
    
    /// The feature layer tracking hurricanes in 2005.
    let featureLayer = AGSFeatureLayer(
        item: AGSPortalItem(
            portal: .arcGISOnline(withLoginRequired: false),
            itemID: "49925d814d7e40fb8fa64864ef62d55e"
        ),
        layerID: 0
    )
    
    /// The time slider from the ArcGIS toolkit.
    let timeSlider = TimeSlider()
    
    /// Configure the time slider's attributes and position.
    func setupTimeSlider() {
        // Configure time slider.
        timeSlider.labelMode = .thumbs
        timeSlider.addTarget(self, action: #selector(timeSliderValueChanged(timeSlider:)), for: .valueChanged)
        // Add the time slider to the view.
        view.addSubview(timeSlider)
        
        // Add constraints to position the slider.
        let margin: CGFloat = 10.0
        timeSlider.translatesAutoresizingMaskIntoConstraints = false
        // Assign the constraints.
        let constraints = [
            timeSlider.bottomAnchor.constraint(equalTo: mapView.attributionTopAnchor, constant: -margin),
            timeSlider.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: margin),
            timeSlider.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -margin)
        ]
        // Activate the constraints.
        NSLayoutConstraint.activate(constraints)
    }
    
    /// Initialize the time slider's steps.
    func initializeTimeStepsFromQuery() {
        featureLayer.load { [weak self] error in
            guard error == nil, let self = self, let timeExtent = self.featureLayer.fullTimeExtent else { return }
            self.populateFeaturesWithQuery {
                self.timeSlider.initializeTimeSteps(timeStepCount: 60, fullExtent: timeExtent) { _ in
                    // Show the time slider.
                    self.timeSlider.playbackInterval = 0.5
                    self.timeSlider.currentExtent = self.currentTimeExtent
                    self.mapView.map!.operationalLayers.add(self.featureLayer)
                }
            }
        }
    }
    
    /// Populate the features using the requested time extent.
    func populateFeaturesWithQuery(completion: @escaping () -> Void) {
        let featureTable = featureLayer.featureTable as! AGSServiceFeatureTable
        // Create query parameters.
        let queryParams = AGSQueryParameters()
        // Set query parameters to filter features based on desired time interval.
        queryParams.timeExtent = currentTimeExtent
        // Set the feature request mode to load the features faster.
        featureTable.featureRequestMode = .manualCache
        // Populate features based on query parameters.
        featureTable.populateFromService(with: queryParams, clearCache: true, outFields: ["*"]) { _, error in
            guard error == nil else { return }
            completion()
        }
    }
    
    @objc
    func timeSliderValueChanged(timeSlider: TimeSlider) {
        if mapView.timeExtent != timeSlider.currentExtent {
            mapView.timeExtent = timeSlider.currentExtent
        }
    }
    
    // MARK: UIViewController
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupTimeSlider()
        initializeTimeStepsFromQuery()
        // Add the source code button item to the right of navigation bar.
        (navigationItem.rightBarButtonItem as? SourceCodeBarButtonItem)?.filenames = ["FilterByTimeExtentViewController"]
    }
}
```

</details>